File: opendpd/core/waveforms/dataset_names.py
"""Compact, deterministic spec summaries; names never define waveform identity."""
def _number(value):
    if 0 < abs(value) < .001:
        mantissa, exponent = f"{value:.2e}".split("e")
        return mantissa.rstrip("0").rstrip(".").replace(".", "p") + "e" + str(int(exponent))
    return f"{value:.3f}".rstrip("0").rstrip(".").replace(".", "p")


def _range(values):
    low, high = min(values), max(values)
    return _number(low) if low == high else f"{_number(low)}-{_number(high)}"


def dataset_name(configs, role="in", model_id=None):
    families = sorted({("nr" if c.preset_id.startswith("nr-") else
                        "w6" if c.preset_id.startswith("wifi6-") else
                        "w7" if c.preset_id.startswith("wifi7-") else c.waveform)
                       if c.waveform == "ofdm" else c.waveform for c in configs})
    parts = ["-".join(families) if len(families) <= 2 else "mix", "bw" + _range([c.bandwidth_hz / 1e6 for c in configs]) + "M"]
    qam = [v for c in configs for v in (c.channel_modulations if c.waveform == "ofdm" else [c.modulation_order] if c.waveform == "qam" else [])]
    psk = [c.psk_order for c in configs if c.waveform == "psk"]
    ofdm = [c for c in configs if c.waveform == "ofdm"]
    if qam:
        parts.append("q" + _range(qam))
    if psk:
        parts.append("p" + _range(psk))
    if ofdm:
        parts.append("c" + _range([len(c.channel_subcarriers) for c in ofdm]))
        spacing = [c.sample_rate_hz / (c.fft_size * c.oversampling * 1000) for c in ofdm]
        if len(set(spacing)) > 1:
            parts.append("s" + _range(spacing) + "k")
    parts.append("n" + str(len(configs)))
    if model_id:
        parts.append(model_id[:20])
    return f"syn_pa_{role}_" + "_".join(parts)

File: frontend/src/utils/datasetNames.ts
import type { GeneratorConfig } from '@/api/signalGenerator'

const number = (value: number) => {
  if (value !== 0 && Math.abs(value) < .001) {
    const [mantissa, exponent] = value.toExponential(2).split('e')
    return mantissa!.replace(/\.?0+$/, '').replace('.', 'p') + 'e' + Number(exponent)
  }
  return value.toFixed(3).replace(/\.?0+$/, '').replace('.', 'p')
}
const range = (values: number[]) => {
  const low = Math.min(...values), high = Math.max(...values)
  return low === high ? number(low) : `${number(low)}-${number(high)}`
}

/** Mirror the server default; immutable signal IDs are independent of these names. */
export function datasetName(configs: GeneratorConfig[], role: 'in' | 'out' | 'inout' = 'in', modelId?: string) {
  if (!configs.length) return `syn_pa_${role}_n0`
  const families = [...new Set(configs.map(c => c.waveform === 'ofdm'
    ? c.preset_id.startsWith('nr-') ? 'nr' : c.preset_id.startsWith('wifi6-') ? 'w6' : c.preset_id.startsWith('wifi7-') ? 'w7' : c.waveform : c.waveform))].sort()
  const parts = [families.length <= 2 ? families.join('-') : 'mix', `bw${range(configs.map(c => c.bandwidth_hz / 1e6))}M`]
  const qam = configs.flatMap(c => c.waveform === 'ofdm' ? c.channel_modulations : c.waveform === 'qam' ? [c.modulation_order] : [])
  const psk = configs.filter(c => c.waveform === 'psk').map(c => c.psk_order)
  const ofdm = configs.filter(c => c.waveform === 'ofdm')
  if (qam.length) parts.push('q' + range(qam))
  if (psk.length) parts.push('p' + range(psk))
  if (ofdm.length) {
    parts.push('c' + range(ofdm.map(c => c.channel_subcarriers.length)))
    const spacing = ofdm.map(c => c.sample_rate_hz / (c.fft_size * c.oversampling * 1000))
    if (new Set(spacing).size > 1) parts.push('s' + range(spacing) + 'k')
  }
  parts.push('n' + configs.length)
  if (modelId) parts.push(modelId.slice(0, 20))
  return `syn_pa_${role}_` + parts.join('_')
}

export const validDatasetName = (name: string, role: 'in' | 'inout') => name.length <= 96 && new RegExp(`^syn_pa_${role}_[A-Za-z0-9][A-Za-z0-9_.-]*$`).test(name)
export function pairedDatasetName(inputName: string, modelId: string) {
  const suffix = '_' + modelId.slice(0, 20)
  return inputName.replace(/^syn_pa_in_/, 'syn_pa_inout_').slice(0, 96 - suffix.length) + suffix
}
